### ISSUE TITLE
feature requests

### ISSUE THREAD

**Issue by guolinke** [microsoft/LightGBM#2302]
This issue is to maintain all features request on one page.

Note to contributors: If you want to work for a requested feature, re-open the linked issue. Everyone is welcome to work on any of the issues below.

Note to maintainers: All feature requests should be consolidated to this page. When there are new feature request issues, close them and create the new entries, with the link to the issues, in this page. 

- [ ] More tests (#261 )
- [ ] Matlab support (#743)
- [ ] String as categorical input directly (#789)
- [ ] AWS S3 support(#1039 )
- [ ] CoreML support(#1074)
- [ ] Parquet file support (#1286)
- [ ] Object importantce (#1460)
- [ ] Include init_score in predict function (#1978)
- [ ] Multiple file as input (#2031)
- [ ] Extracting decision path(#2187)

new algorithm:
- [ ] Regularized Greedy Forest (#315)
- [ ] Accelerated Gradient Boosting (#1257)
- [ ] Piece-wise linear tree(#1315)
- [ ] Multi layer gradient boosting tree(#1423)
- [ ] Adaptive neural tree(#1542) 
- [ ] Probabilistic Random Forest(#1946)
- [ ] Sparrow(#2001)

objective and metric:
- [ ] Multi-output regression (#524)
- [ ] Earth Mover Distance (#1256)
- [ ] Cox Proportional Hazard Regression (#1837)
- [ ] Ranking metric for regression objective (#1911)
- [ ] Density estimation (#2056)

python package:
- [ ] Keep cv predicted values (#283)
- [ ] Python input checking (#812)
- [ ] Log redirect in python (#1493)

R package:
- [ ] Yarn support (#790)
- [ ] Plotting in R-package(#1222)
- [ ] H2O datatable support (#2003)
- [ ] [R-package] Allow data to live in memory mapped file (#2148)

GPU:
- [ ] Multiple GPU support (#620)
- [ ] GPU performance improvement (#768)
- [ ] GPU binarines release (#2263)

**Comment by MotoRZR** [comment]
For everyone's information, the sparrow algorithm has been implemented in CatBoost

**Comment by MotoRZR** [comment]
There’s a reference to minimum variance sampling here:

https://catboost.ai/docs/concepts/algorithm-main-stages_bootstrap-options.html

Although I think it just speeds up training rather than providing out of core training.

**Comment by momijiame** [comment]
I would like to tackle the following issues on Python package. Could I discuss about a plan to fix? Also, where can we discuss that? IMHO, They will be resolved by improving to ``lightgbm.cv()`` function.

#2105: Make _CVBooster public for better stacking experience
#283: Keep cv predicted values

I want to reopen the above issues, but I can not do that. Maybe I have no permission.

**Comment by guolinke** [comment]
we would like to call the voting here, to prioritize these requests.
If you think a feature request is very necessary for you, you can vote for it by the following process:

1. got the issue (feature request) number.  
2. search the number in this issue, check the voting of it exists or not.
3. if the voting exists, you can add 👍 to that voting
4. if the voting doesn't exist, you can create a new voting by replying to this thread, and add the number in the it.

**Comment by StrikerRUS** [comment]
> we would like to call the voting here

Let me start.


#2644

**Comment by candalfigomoro** [comment]
It was proposed by me so I'm a little bit biased

#3128

**Comment by candalfigomoro** [comment]
GPU binaries release #2263

**Comment by candalfigomoro** [comment]
Enhance parameter tuning guide with more params #2617

**Comment by candalfigomoro** [comment]
Subsampling rows with replacement #1038

**Comment by candalfigomoro** [comment]
Piece-wise linear tree #1315 (also see PR https://github.com/microsoft/LightGBM/pull/3299)

**Comment by candalfigomoro** [comment]
Multi-output regression #524

**Comment by candalfigomoro** [comment]
Cox Proportional Hazard Regression #1837

**Comment by jameslamb** [comment]
Based on https://github.com/microsoft/LightGBM/issues/2983#issuecomment-722630931, I've updated this issue's description:

> Note to **maintainers**: All feature requests should be consolidated to this page. When there are new feature request issues, close them and create the new entries, with the link to the issues, in this page. The one exception is issues marked `good first issue`...these should be left open so they are discoverable by new contributors.

I think that we should keep `good first issue` issues  open, so it's easy for new contributors to find them.

**Comment by gzls90** [comment]
Read from multiple files and parquet support #2031

**Comment by gzls90** [comment]
Parquet file support #1286

**Comment by gzls90** [comment]
Register custom objective / loss function #3244

**Comment by gzls90** [comment]
Object importantce #1460

**Comment by gzls90** [comment]
Support trees with linear models at leaves #3319

**Comment by wenmin-wu** [comment]
read from multiple zipped libsvm format text files

**Comment by gzls90** [comment]
Multiple GPU support [#620 ](https://github.com/microsoft/LightGBM/issues/620)

**Comment by 7starsea** [comment]
Multiple GPU support (#620)
(From my experience, the xgboost with gpu seems faster than lightgbm with gpu.)

**Comment by GLevV** [comment]
1. Native support for Focal Loss. There are some implementations for LGBM and XGB. They show better performance on imbalanced data, but implementations are slow and clunky. Also allow to use MSE for classification (also known as [square loss](https://en.wikipedia.org/wiki/Loss_functions_for_classification#Square_loss)). More classification losses could be added to stand out in this department (since all GBDTs algorithms use LogLoss and nothing else).
2. Different quantization techniques. As I understand most of GBDTs use quantiles to quantize data (like pd.qcut or KBinsDiscretizer(strategy=quantile), but there are other options, like uniform quantization (like pd.cut or KBinsDiscretizer(strategy=uniform). CB has a bunch of other techniques, but for starter these two would be just fine.

**Comment by StrikerRUS** [comment]
For everyone who was voting for multi-gpu support, please try our new experimental CUDA version which was kindly contributed by our friends from IBM. This version supports multi-GPU training. We will really appreciate any early feedback on this experimental feature (please create new issues, do not comment here).

How to install: https://lightgbm.readthedocs.io/en/latest/Installation-Guide.html#build-cuda-version-experimental.

Argument to specify number of GPUs: https://lightgbm.readthedocs.io/en/latest/Parameters.html#num_gpu.

**Comment by jz457365** [comment]
Support ignoring some features during training on constructed dataset #4317

**Comment by jz457365** [comment]
Spike and slab feature sampling priors (feature weighted sampling) #2542

**Comment by bethrice44** [comment]
Quantile LightGBM: ensure monotonic [#3447](https://github.com/microsoft/LightGBM/issues/3447)

**Comment by robo-sq** [comment]
SHAP feature contribution for linear trees #4002

**Comment by ira-saktor** [comment]
Create dataset from pyarrow tables: #3369

**Comment by js850** [comment]
Add support for CRLF line endings or improve documentation and error message #5508

**Comment by thomaslundgaard** [comment]
[Optimisations for Apple Silicon #3606](https://github.com/microsoft/LightGBM/issues/3606)

**Comment by antaradas94** [comment]
Add parameter to control maximum group size for Lambdarank  #5053

**Comment by chopeen** [comment]
Allow training without loading full dataset into memory #5094

**Comment by chopeen** [comment]
Support different data types (when load data from Python) #3459

**Comment by szjunma** [comment]
Add support for early stopping in Dask interface #3712

**Comment by vitorpbarbosa7** [comment]
Add Earth Mover Distance as objective metric to be optimized (maximized) #1256

**Comment by tim-habitat** [comment]
Apache Arrow seems to be gaining a lot of traction in the dataframe space. 
We use `polars` and it would be great to be able to directly create a dataset from `arrow` format.
Also, `pandas 2.0` will have `arrow` as a backend later this month .

**Comment by barynton** [comment]
Conan installation support #5770

**Comment by sanurielf** [comment]
Add support for Multi-output regression #524

**Comment by vladv14** [comment]
Provide access to the bin ids and bin upper bounds of the constructed dataset #5191

**Comment by onacrame** [comment]
Consider implementation of the sketchboost algorithm for multi output/multiclass setting.  The current multiclass approach is highly ineffecient as a separate tree structure is required for each class.  This approach significantly improves on training time and model size by allowing a single tree structure to handle many classes.  

This is already implemented in the Py-Boost library.

https://arxiv.org/pdf/2211.12858.pdf

**Comment by kaizhu256** [comment]
WebAssembly support (https://github.com/microsoft/LightGBM/issues/5372)

**Comment by jane-delaney** [comment]
Support monotone constraints with quantile objective #3371

**Comment by bhvieira** [comment]
Recalculate feature importance during the update process of a tree model / Calculate Gain Importance on Test Data (#2413)

**Comment by cgoo4** [comment]
Add R-package support for an early-stopping `min_delta` as implemented in Python #4580.

**Comment by humairan** [comment]
Add support for CRLF line endings or improve documentation and error message #5508

**Comment by pengxiao-song** [comment]
Allow specifying probability measure for features #4605

**Comment by dkaufman-rc** [comment]
Support ignoring some features during training on constructed dataset [#4317](https://github.com/microsoft/LightGBM/issues/4317)

**Comment by hsorsky** [comment]
#2921 Better document missing values behavior during prediction (https://github.com/microsoft/LightGBM/issues/2921)

**Comment by nucflash** [comment]
Add ability to predict on `Dataset` (https://github.com/microsoft/LightGBM/issues/2666, https://github.com/microsoft/LightGBM/issues/6613, https://github.com/microsoft/LightGBM/issues/6285)

**Comment by kszlim** [comment]
Allow JSON special characters in feature names (https://github.com/microsoft/LightGBM/issues/6202)

**Comment by jihadjh808-dot** [comment]
# About searching on GitHub

Our integrated search covers the many repositories, users, and lines of code on GitHub.

## About searching on GitHub

You can search globally across all of GitHub, or scope your search to a particular repository or organization.

* To search globally across all of GitHub, type what you're looking for into the search field at the top of any page, and choose "Search all of GitHub" in the search dropdown menu.

* To search within a particular repository or organization, navigate to the repository or organization page, type what you're looking for into the search field at the top of the page, and press **Enter**.

  You can also use suggestions and completions in the search bar to quickly find what you need.

* If you click on the search bar in the top navigation of GitHub.com, you will see a list of suggestions organized by category, including recent searches and suggested repositories, teams, and projects that you have access to.

* Clicking on any of the specific suggestions will take you directly to the page for that suggestion (for example, the repository or project page). If you click on a recent search, depending on the type of search, the search term will appear in the search bar or you will be taken to the search results page for the search term.

* Once you start typing, you will see a list of completions and suggestions that match your query. You can click on a suggestion to jump to a specific location. As you continue to type, you will see more specific suggestions, such as code files you can jump to directly.

After typing a search query, you can press **Enter** to go to the full search results view, where you can see each match and a visual interface for applying filters. For more information, see [Searching using a visual interface](#searching-using-a-visual-interface).

> \[!NOTE]
>
> * You must be signed into a personal account on GitHub to search for code across all public repositories.
> * GitHub Pages sites are not searchable on GitHub. However, you can search the source content if it exists in the default branch of a repository, using code search. For more information, see [Understanding GitHub Code Search syntax](/en/search-github/github-code-search/understanding-github-code-search-syntax). For more information about GitHub Pages, see [What is GitHub Pages?](/en/pages/getting-started-with-github-pages/what-is-github-pages).
> * Currently our search doesn't support exact matching.

After running a search on GitHub, you can sort the results, or further refine them by clicking one of the languages in the sidebar. For more information, see [Sorting search results](/en/search-github/getting-started-with-searching-on-github/sorting-search-results).

GitHub search uses an ElasticSearch cluster to index projects every time a change is pushed to GitHub. Issues and pull requests are indexed when they are created or modified.

## Types of searches on GitHub

You can search for the following information across all repositories you can access on GitHub.

* [Repositories](/en/search-github/searching-on-github/searching-for-repositories)
* [Topics](/en/search-github/searching-on-github/searching-topics)
* [Issues and pull requests](/en/search-github/searching-on-github/searching-issues-and-pull-requests)
* [Discussions](/en/search-github/searching-on-github/searching-discussions)
* [Code](/en/search-github/github-code-search/understanding-github-code-search-syntax)
* [Commits](/en/search-github/searching-on-github/searching-commits)
* [Users](/en/search-github/searching-on-github/searching-users)
* [Packages](/en/search-github/searching-on-github/searching-for-packages)
* [Wikis](/en/search-github/searching-on-github/searching-wikis)

## Searching using a visual interface

In addition to the search bar, you can search GitHub using the [search](https://github.com/search) page or [advanced search](https://github.com/search/advanced) page. Alternatively, you can use the interactive search in the GitHub Command Palette to search your current location in the UI, a specific user, repository or organization, and globally across all of GitHub, without leaving the keyboard. For more information, see [GitHub Command Palette](/en/get-started/accessibility/github-command-palette).

The [advanced search](https://github.com/search/advanced) page provides a visual interface for constructing search queries. You can filter your searches by a variety of factors, such as the number of stars or number of forks a repository has. As you fill in the advanced search fields, your query will automatically be constructed in the top search bar.

![Advanced Search page. Top search bar holds "kittens user:octocat" query. Under "Advanced options", "From these owners" text box holds term "octocat".](/assets/images/help/search/advanced-search.png)

## Searching repositories on GitHub.com from your private enterprise environment

If you use GitHub.com or GHE.com as well as GitHub Enterprise Server, and an enterprise owner has enabled unified search, you can search across both environments at the same time from GitHub Enterprise Server. For more information, see [About searching on GitHub](/en/enterprise-server@3.22/search-github/getting-started-with-searching-on-github/about-searching-on-github#searching-repositories-on-githubcom-from-your-private-enterprise-environment) in the GitHub Enterprise Server documentation.

## Further reading

* [Understanding the search syntax](/en/search-github/getting-started-with-searching-on-github/understanding-the-search-syntax)
* [Searching on GitHub](/en/search-github/searching-on-github)

**Comment by jihadjh808-dot** [comment]
# About searching on GitHub

Our integrated search covers the many repositories, users, and lines of code on GitHub.

## About searching on GitHub

You can search globally across all of GitHub, or scope your search to a particular repository or organization.

* To search globally across all of GitHub, type what you're looking for into the search field at the top of any page, and choose "Search all of GitHub" in the search dropdown menu.

* To search within a particular repository or organization, navigate to the repository or organization page, type what you're looking for into the search field at the top of the page, and press **Enter**.

  You can also use suggestions and completions in the search bar to quickly find what you need.

* If you click on the search bar in the top navigation of GitHub.com, you will see a list of suggestions organized by category, including recent searches and suggested repositories, teams, and projects that you have access to.

* Clicking on any of the specific suggestions will take you directly to the page for that suggestion (for example, the repository or project page). If you click on a recent search, depending on the type of search, the search term will appear in the search bar or you will be taken to the search results page for the search term.

* Once you start typing, you will see a list of completions and suggestions that match your query. You can click on a suggestion to jump to a specific location. As you continue to type, you will see more specific suggestions, such as code files you can jump to directly.

After typing a search query, you can press **Enter** to go to the full search results view, where you can see each match and a visual interface for applying filters. For more information, see [Searching using a visual interface](#searching-using-a-visual-interface).

> \[!NOTE]
>
> * You must be signed into a personal account on GitHub to search for code across all public repositories.
> * GitHub Pages sites are not searchable on GitHub. However, you can search the source content if it exists in the default branch of a repository, using code search. For more information, see [Understanding GitHub Code Search syntax](/en/search-github/github-code-search/understanding-github-code-search-syntax). For more information about GitHub Pages, see [What is GitHub Pages?](/en/pages/getting-started-with-github-pages/what-is-github-pages).
> * Currently our search doesn't support exact matching.

After running a search on GitHub, you can sort the results, or further refine them by clicking one of the languages in the sidebar. For more information, see [Sorting search results](/en/search-github/getting-started-with-searching-on-github/sorting-search-results).

GitHub search uses an ElasticSearch cluster to index projects every time a change is pushed to GitHub. Issues and pull requests are indexed when they are created or modified.

## Types of searches on GitHub

You can search for the following information across all repositories you can access on GitHub.

* [Repositories](/en/search-github/searching-on-github/searching-for-repositories)
* [Topics](/en/search-github/searching-on-github/searching-topics)
* [Issues and pull requests](/en/search-github/searching-on-github/searching-issues-and-pull-requests)
* [Discussions](/en/search-github/searching-on-github/searching-discussions)
* [Code](/en/search-github/github-code-search/understanding-github-code-search-syntax)
* [Commits](/en/search-github/searching-on-github/searching-commits)
* [Users](/en/search-github/searching-on-github/searching-users)
* [Packages](/en/search-github/searching-on-github/searching-for-packages)
* [Wikis](/en/search-github/searching-on-github/searching-wikis)

## Searching using a visual interface

In addition to the search bar, you can search GitHub using the [search](https://github.com/search) page or [advanced search](https://github.com/search/advanced) page. Alternatively, you can use the interactive search in the GitHub Command Palette to search your current location in the UI, a specific user, repository or organization, and globally across all of GitHub, without leaving the keyboard. For more information, see [GitHub Command Palette](/en/get-started/accessibility/github-command-palette).

The [advanced search](https://github.com/search/advanced) page provides a visual interface for constructing search queries. You can filter your searches by a variety of factors, such as the number of stars or number of forks a repository has. As you fill in the advanced search fields, your query will automatically be constructed in the top search bar.

![Advanced Search page. Top search bar holds "kittens user:octocat" query. Under "Advanced options", "From these owners" text box holds term "octocat".](/assets/images/help/search/advanced-search.png)

## Searching repositories on GitHub.com from your private enterprise environment

If you use GitHub.com or GHE.com as well as GitHub Enterprise Server, and an enterprise owner has enabled unified search, you can search across both environments at the same time from GitHub Enterprise Server. For more information, see [About searching on GitHub](/en/enterprise-server@3.22/search-github/getting-started-with-searching-on-github/about-searching-on-github#searching-repositories-on-githubcom-from-your-private-enterprise-environment) in the GitHub Enterprise Server documentation.

## Further reading

* [Understanding the search syntax](/en/search-github/getting-started-with-searching-on-github/understanding-the-search-syntax)
* [Searching on GitHub](/en/search-github/searching-on-github)
